Adds --manifest-url flag to allow plugin installation from remote manifests

rewanth1997 opened the pull request:
This feature addition lets the user/developer to install a krew plugin directly from the source URL instead of manually downloading the plugin file.

## Problem statement
Install kubectl plugin from config file

### Current approach
```
wget https://raw.githubusercontent.com/rewanth1997/kubectl-fields/master/deploy/krew/plugin.yaml
kubectl krew install --manifest=./plugin.yaml
```

### Proposed approach
```
kubectl krew install --manifest-url="https://github.com/rewanth1997/kubectl-fields/raw/master/deploy/krew/plugin.yaml"
```

This new feature addition will download the plugin from the URL to a temporary location and then installs the plugin from that temporary fie using the regular `manifest` method.

Please review the feature request. I'm happy to make the necessary changes if any.

Regards,
Rewanth

rewanth1997 commented:
/assign @soltysh

ahmetb commented:
This proposal was previously discussed & denied, I believe.

rewanth1997 commented:
Oh, I couldn't find any such discussion on Github. 

This PR downloads the yaml file from the URL and stores it in a temporary location. After download, it performs the same operation as the `--manifest` flag. After installation, this feature cleans up the file from the system.

I couldn't think of any potential drawbacks. In fact, I see this as an advantage for developers/users. May I know the reason behind denying this idea?

ahmetb commented:
I also tried to find but couldn't. Now I'm thinking if it was a private discussion with @corneliusweig.

The main reasoning was a about a few points:

1. You can just do `--manifest <(curl url)` and compose tools/pipes/sub-shells as one should do with UNIX command-line tools.

2. `--manifest-url` is yet another thing to maintain, and makes the user/test surface larger. We thought about detecting url in `--manifest` but that is also "magic", which is not a good thing to do in tools.

Since this has came up multiple times by now, we might be open to it.

/unassign @soltysh 
/assign
/assign @corneliusweig

corneliusweig commented:
Hi @rewanth1997, the issue @ahmetb mentioned is #193.

---

I think the main reason against the `--manifest-url` option is that this is only intended for plugin devs and not for krew users. Adding this `--manifest-url` option would also encourage end-users to install dangling plugins via this option. However, we don't want this, because `krew` should also take care of updates and this doesn't work when installing from URL.

However, even plugin devs do stumble over this `--manifest <(curl URL)` thing. How about this: if a user tries to install with `--manifest URL`, we could detect if the argument is in fact a URL and print an error suggesting to try `--manifest <(curl URL)` instead?

rewanth1997 commented:
I wasn't aware of the `update` issue.

I just check the `update` module in krew, https://github.com/kubernetes-sigs/krew/blob/master/pkg/gitutil/git.go#L59. With this implementation, the plugins installed via `kubectl install --manifest <(curl ...)` won't be updated.

The idea of throwing a custom error to use curl seems good but building a generic regex expression for URL is very tough. Limiting this just to `http/https` protocol check isn't a good idea either.

corneliusweig commented:
@rewanth1997 Implementing the URL parser is not necessary. Go already has that built in. See https://github.com/kubernetes-sigs/krew/pull/302/files
```go
if _, err := url.ParseRequestURI(word); err == nil {
   return true
}
```

rewanth1997 commented:
I don't think this works properly,

`url.ParseRequestURI` considers anything starting with "/" as URL. It might be an absolute path to the plugin file in the host machine.

"/home/rewanth/kubectl-fields/deploy/krew/plugin.yaml" gets detected as URL.

```
package main

import (
    "fmt"
    "net/url"
)

func main() {
    fmt.Println(isUrl("https://github.com/kubernetes-sigs/krew/pull/350"))
    fmt.Println(isUrl("/home/rewanth/kubectl-fields/deploy/krew/plugin.yaml"))
    fmt.Println(isUrl("plugin.yaml"))
}

func isUrl(toTest string) bool {
    _, err := url.ParseRequestURI(toTest)
    if err != nil {
        return false
    } else {
        return true
    }
}
```

Output
```
true
true
false
```

rewanth1997 commented:
```
rewanth@ubuntu:/tmp/krew/out/bin$ ./krew-linux_amd64 install --manifest=https://raw.githubusercontent.com/rewanth1997/kubectl-fields/master/deploy/krew/plugin.yaml
F1013 22:59:14.313896  107113 root.go:58] Cannot install plugin directly from URL
To install from URL, use --manifest <(curl https://raw.githubusercontent.com/rewanth1997/kubectl-fields/master/deploy/krew/plugin.yaml)
```

@corneliusweig I added an additional check to check for `Schema` and `Host` to validate the URL. Please review the PR.

corneliusweig commented:
Good find! I'll consider adding the same logic in the other PR.

Can you also add some tests for this behavior? Otherwise this looks good. Thanks!

rewanth1997 commented:
I validated the module independently with different protocols like SMB, FTP, HTTP, etc. I believe these checks are good enough for integration. Please let me know if something else needs to be done.

List of tests performed: [https://play.golang.org/p/xUN0SHp1c9t](https://play.golang.org/p/xUN0SHp1c9t)

ahmetb commented:
@corneliusweig have we settled on --manifest vs --manifest-url?
It seems like latter is more verbose (as opposed to _magic_) and can be easily implemented, albeit with extra testing requirements.

corneliusweig commented:
@ahmetb The scope of this PR was changed. It now only prints an error message with a hint to use `--manifest <(curl filename)`, if it detects that `filename` is a remote URL.

corneliusweig commented:
/retitle Print hint when user tries to install via manifest from remote URL

corneliusweig commented:
@rewanth1997 You could add an e2e test for install with manifest which does one of the checks automatically (e.g. `https://github.com/kubernetes-sigs/krew/pull/350`). A good place would be in `integration_test/install_test.go`.

rewanth1997 commented:
Hi @corneliusweig @ahmetb ,

I have a concern about the current idea. You mentioned in the above [comment-539180993](https://github.com/kubernetes-sigs/krew/pull/350#issuecomment-539180993) that `--manifest-url` has to be avoided because it installs dangling plugins and `krew update` is not applied on them. In that case, even the plugins installed by `--manifest` argument cannot be updated by `krew update`.

If `--manifest` is allowed to install dangling plugins, I think `--manifest-url` should also be allowed.

**Proposed solution:**
We will revert back to my original commit of adding `--manifest-url` and add a warning like *"Plugins installed via --manifest/--manifest-url are not updated by krew"* when the users install plugin with `--manifest/--manifest-url`

What are your views on this new proposal?

ahmetb commented:
Yeah I think --manifest-url is not that bad.
We just need to write tests for it (for every case we've been testing with --manifest).

No need to add a warning. The flag description clearly says `(Development-only)`.

rewanth1997 commented:
I already implemented the code for `--manifest-url`. I can add the test cases too for this module.

Once we get @corneliusweig views on adding `--manifest-url`, I will revert back to first commit and will add the test cases as well.

corneliusweig commented:
@rewanth1997 Alright then, let's have the `--manifest-url` flag. Sorry if I caused additional work with this back-and-forth. Please also change back the PR title to something meaningful.

rewanth1997 commented:
Not a problem. I will make the required changes.

rewanth1997 commented:
/retitle Adds --manifest-url option to allow installation from remote manifests

@corneliusweig Please review the latest code which includes integration tests as well

rewanth1997 commented:
@corneliusweig Your kconfig repo (https://github.com/corneliusweig/konfig/) isn't having a valid plugin structured yaml file which can be passed to `--manifest-url` option for running the integration tests.

Hence, I added my plugin file (kubectl-fields) to run the integration tests.

corneliusweig commented on cmd/krew/cmd/install.go:
```diff
@@ -62,6 +99,20 @@ Remarks:
+			if *manifestURL != "" {
+				fileName, err := downloadFile(*manifestURL)
+				if err != nil {
+					fmt.Fprintln(os.Stderr, "Error downloading file")
```
Please do
```go
return errors.Wrapf(err, "Error downloading manifest from %q", *manifestURL)
```

Just logging an error won't be enough ;)

---

Please also put some thought into the question whether the cleanup function (remove the temp-file) should be called when an error is returned here. Otherwise we might leak temporary files.

rewanth1997 replied:
Done. I made a call to cleanup function to clear the temp files.

corneliusweig commented on cmd/krew/cmd/install.go:
```diff
@@ -30,10 +32,45 @@ import (
 )
 
+// Downloads data from url to a filepath location
+func downloadFile(url string) (string, error) {
```
Can you move that function into the `internal` package?

rewanth1997 replied:
Done

corneliusweig commented on cmd/krew/cmd/install.go:
```diff
@@ -62,6 +99,20 @@ Remarks:
 			copy(pluginNames, args)
 
+			// Downloads manifest file from given URL
+			if *manifestURL != "" {
```
I wonder if you could find a way to not make the `install` function even longer. It already is quite long and is very hard to test. Adding this here only makes it worse. So if you find a way to move this code into some helper where it can be tested independently, that would be great!

rewanth1997 replied:
Unfortunately no. We cannot shift this piece of code to any helper function because we use `defer` to cleanup the temporary files after installation here (https://github.com/rewanth1997/krew/blob/manifest-url/cmd/krew/cmd/install.go#L74)

If we move the piece of code to another function, `defer` will work locally to the particular helper function only and will loose its primary purpose.

rewanth1997 replied:
I tried to make the code as minimal as possible in this section. Its only 5 lines of code.
```
downloadFile()
cleanup()
check_err
assign temporary file name to the manifest variable
```

ahmetb replied:
Sorry I'm jumping on this but why do we need to write the temporary file onto the disk? 

Maybe if we changed the installation parameter with an `io.ReadCloser` we can solve this more cleanly?

rewanth1997 replied:
My initial idea was to the downloaded data to a temporary file on the disk and then to pass the temporary file path as a value to `--manifest` variable.

Thanks for pointing the `ReadCloser` interface. This will definitely make the code look cleaner. I can directly save the downloaded file in a variable/buffer and pass it to respective variables rather than storing on disk and removing it later. I will make the changes asap.

corneliusweig commented on pkg/constants/constants.go:
```diff
@@ -19,6 +19,7 @@ const (
 	PluginKind        = "Plugin"
 	ManifestExtension = ".yaml"
 	KrewPluginName    = "krew" // plugin name of krew itself
+	ManifestURL       = "https://github.com/rewanth1997/kubectl-fields/raw/master/deploy/krew/plugin.yaml"
```
These constants are for production code only. I think `testutil_test.go` would be a better place for this.

rewanth1997 replied:
Done

corneliusweig commented on integration_test/install_test.go:
```diff
@@ -105,6 +105,17 @@ func TestKrewInstall_Manifest(t *testing.T) {
 	test.AssertExecutableInPATH("kubectl-" + validPlugin)
 }
 
+func TestKrewInstall_ManifestURL(t *testing.T) {
```
Maybe also add a test-case where `--manifest-url` _and_ `--archive` are given? What should happen in that case?

ahmetb replied:
I think we can make these flags mutually exclusive and explicitly error out. People most probably won't even use it.

That said, --archive just always uses the provided file instead of downloading the .zip file from the internet. So it's actually still valid here, but that's not why people are going to use --manifest-url for. 

So yeah, I recommend 1) explicitly fail in this case 2) add integration tests to test that.

rewanth1997 replied:
Done

rewanth1997 commented on cmd/krew/cmd/install.go:
```diff
@@ -30,10 +32,45 @@ import (
 )
 
+// Downloads data from url to a filepath location
+func downloadFile(url string) (string, error) {
```
Done

rewanth1997 commented on cmd/krew/cmd/install.go:
```diff
@@ -30,10 +32,45 @@ import (
+		return "", fmt.Errorf("bad status: %s", resp.Status)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
```
Done

rewanth1997 commented:
Hi @corneliusweig ,

Apologies for the delay. I made all the recommended changes, please review the latest code.

```
Do you think you could set up a local test server with import "net/http/httptest" to serve the manifest for tests?
```
I didn't understand this question. Can you elaborate more?

rewanth1997 commented:
Sorry for the delay @corneliusweig @ahmetb 
I was busy working on a client's project and didn't get time to fix this PR. Since I have free time now, I will be more active and resolve the issues quickly.

corneliusweig commented:
@rewanth1997 What I mean is that you don't fetch the manifest from some real repo, but instead you create an http server that serves the manifest and the archive. This has two advantages:

- When working on krew, we have full control over the manifest and archive. If the test breaks, we can fix it.
- There will be a much smaller delay when "downloading", because the files are served locally. So the tests will run faster.

I think the `"net/http/httptest"` package has the required utils to do this.

ahmetb commented:
Some examples with httptest pkg:
- https://itnext.io/how-to-stub-requests-to-remote-hosts-with-go-6c2c1db32bf2
- https://medium.com/what-i-talk-about-when-i-talk-about-technology/go-code-examples-httptest-newserver-f965fb349884
- https://medium.com/@xoen/go-testing-technique-testing-json-http-requests-76d9ce0e11f

rewanth1997 commented:
@ahmetb @corneliusweig  I stored the downloaded file in the buffer. Please share your views on the latest implementation.

I will add test cases getting served via local server asap.

rewanth1997 commented:
@ahmetb @corneliusweig All the recommended changes have been done. Please review the code that serves files from the local test server.

ahmetb commented on cmd/krew/cmd/install.go:
```diff
@@ -94,6 +102,21 @@ Remarks:
+				if err := validation.ValidatePlugin(plugin.Name, plugin); err != nil {
+					return errors.Wrap(err, "plugin manifest validation error")
+				}
+				install = append(install, plugin)
```
this is too repetitive with the next block, it's not good for code cleanliness.

rewanth1997 replied:
True. 

he only alternative is to define `plugin` and `err` as global variables and then pass them to the repetitive block of code. I thought it would be a bad idea to declare variables with global scope and I preferred the repetitive method.

I can implement the alternative method as explained above if that makes sense.

ahmetb replied:
hmm, why don't we make `decodepluginfile` work on io.ReadCloser?

Both resp.Body and os.OpenFile return io.ReadCloser.

rewanth1997 replied:
```
var plugin index.Plugin
var gerr error

if *manifestURL != "" {
	resp, err := internal.DownloadFile(*manifestURL)
	if err != nil {
		return errors.Wrapf(err, "Error downloading manifest from %q", *manifestURL)
	}
	plugin, gerr = indexscanner.DecodePluginFile(resp)
} else if *manifest != "" {
	plugin, gerr = indexscanner.ReadPluginFile(*manifest)
}

if gerr != nil {
	return errors.Wrap(gerr, "failed to load custom manifest file")
}
if err := validation.ValidatePlugin(plugin.Name, plugin); err != nil {
	return errors.Wrap(err, "plugin manifest validation error")
}
install = append(install, plugin)
```

How does this look? This snippet will replace this piece of code, https://github.com/kubernetes-sigs/krew/blob/6df65f26ed4b6420ad8ed0678ddb4cfe0a0ad99e/cmd/krew/cmd/install.go#L105-L129

ahmetb replied:
good enough but instead of creating `gerr`, you should probably wrap/return errors immediately where they occur.

Ideally I'd love to see us using `indexscanner.Decode` just accept `io.ReadCloser`. Maybe we can do it later, not in this PR.

rewanth1997 replied:
Done

ahmetb commented on integration_test/testdata/konfig_localhost.yaml:
```diff
@@ -0,0 +1,28 @@
+apiVersion: krew.googlecontainertools.github.com/v1alpha2
```
why do we need this file?
As far as I can tell you don't need to fetch the plugin from a hardcoded port number.
All you care about while implementing `--manifest-url` is that manifest is read from URL and then you can pass `--archive` override.

you put `uri: http://ignored`

rewanth1997 replied:
I added a testcase in which `--archive` is not specified along with `--manifest-url`. How can we ignore the URL in that case? We would need a valid URL to download the plugin.

rewanth1997 replied:
Done

ahmetb commented on integration_test/testutil_test.go:
```diff
@@ -37,8 +37,10 @@ import (
-	validPlugin2         = "mtail"  // a plugin in central index with small size
+	validPlugin          = "konfig"          // a plugin in central index with small size
+	validPlugin2         = "mtail"           // a plugin in central index with small size
+	LocalhostURL         = "127.0.0.1:60564" // make sure this port is open. incase of editing port number, make sure to update the URL in testdata/konfig_localhost.yaml
```
you shouldn't need this. just simplify by reading tarball from `--archive` in integration test?

rewanth1997 replied:
As we discussed previously `--archive` cannot be used with `--manifest-url`. Hence, I added this.

ahmetb commented on integration_test/install_test.go:
```diff
@@ -105,6 +111,95 @@ func TestKrewInstall_Manifest(t *testing.T) {
+
+	server, err := startServer()
+	if err != nil {
+		t.Errorf("Trouble starting local server")
```
you aren't logging what the error was. just t.Fatalf(error)

Also errorf doesn't stop the test. so the test will continue to execute.

rewanth1997 replied:
Done

ahmetb commented on integration_test/install_test.go:
```diff
@@ -105,6 +111,95 @@ func TestKrewInstall_Manifest(t *testing.T) {
+		"--manifest-url", LocalhostManifestURL,
+		"--manifest", filepath.Join("testdata", fooPlugin+constants.ManifestExtension)).Run()
+	if err == nil {
+		t.Errorf("Expected install to fail but was successful")
```
lowercase errors

rewanth1997 replied:
Done

ahmetb commented on integration_test/install_test.go:
```diff
@@ -105,6 +111,95 @@ func TestKrewInstall_Manifest(t *testing.T) {
+func startServer() (*httptest.Server, error) {
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.RequestURI {
+		case "/konfig_localhost.yaml":
```
why don't you just use http.FileServer ?

rewanth1997 replied:
I thought having a custom handler function will be helpful if we need to serve additional files in the future from different directories.

ahmetb replied:
But we currently don’t, and this implementations is 20x longer. :)

rewanth1997 replied:
I will fix it

rewanth1997 replied:
Done

ahmetb commented:
Currently this PR is too big. Ideally you should add 1 if statement + 1 unit test to cover that the parsing happens. Nothing else should be necessary as far as I can tell, overall please try to reduce complexity. We shouldn't need ~200 lines of code for a simple flag.

rewanth1997 commented:
I agree that this PR is a bit lengthy.

The recommended changes are made and the size of this PR is notably reduced. I added extra unit tests with a comination of `--manifest-url` and other flags like `--archive` and `--manifest`.

This PR has 1 if statement, a file that contains code to download files from URL, 3 unit test cases and the setup of a local test server. I think ~150 lines of code is a reasonable length to code all the above operations.

@ahmetb @corneliusweig Please review the latest code.

ahmetb commented on integration_test/install_test.go:
```diff
@@ -105,6 +110,75 @@ func TestKrewInstall_Manifest(t *testing.T) {
+	skipShort(t)
+
+	server, err := startLocalServer()
+	if err != nil {
```
Do this:

Pass t to startLocalServer, Mark that method as t.Helper() and if err starting server, t.Fatal in that method; not here.

rewanth1997 replied:
Done

ahmetb commented on integration_test/testutil_test.go:
```diff
@@ -37,8 +37,10 @@ import (
-	validPlugin2         = "mtail"  // a plugin in central index with small size
+	validPlugin          = "konfig"          // a plugin in central index with small size
+	validPlugin2         = "mtail"           // a plugin in central index with small size
+	localServerURL       = "127.0.0.1:45076" // change port number incase of bind error
```
Please do not hardcode. 

httputil can start a server at a random port. 
Just return its base url and close function from test util func.

rewanth1997 replied:
Done

corneliusweig commented on cmd/krew/cmd/install.go:
```diff
@@ -94,11 +102,27 @@ Remarks:
+					plugin, err = indexscanner.DecodePluginFile(resp)
+					if err != nil {
+						return errors.Wrap(err, "failed to load custom manifest file")
+					}
```
this should go into `internal.DownloadFile`

rewanth1997 replied:
Done

corneliusweig commented on cmd/krew/cmd/internal/downloadFile.go:
```diff
@@ -0,0 +1,30 @@
+
+// DownloadFile function takes URL as input, stores the contents in temporary file and returns filename, cleanup function to delete temporary files and error
+func DownloadFile(url string) (io.ReadCloser, error) {
+	resp, err := http.Get(url)
```
`http.Get` returns a response object whose `Body` is expected to be closed by the user. See the documentation for `http.Do` for details:

> // If the returned error is nil, the Response will contain a non-nil
// Body which the user is expected to close. If the Body is not both
// read to EOF and closed, the Client's underlying RoundTripper
// (typically Transport) may not be able to re-use a persistent TCP
// connection to the server for a subsequent "keep-alive" request.

I think it would make more sense to not just fetch any file with this method, but instead move the whole plugin parsing (`indexscanner.DecodePluginFile`) into this function (see the previous comment). This also improves the testability.

Consequently, the method should be called `ReadRemotePluginManifest` or similar.

rewanth1997 replied:
Done

corneliusweig commented on cmd/krew/cmd/install.go:
```diff
@@ -62,6 +62,14 @@ Remarks:
 			var pluginNames = make([]string, len(args))
 			copy(pluginNames, args)
 
+			if *archiveFileOverride != "" && *manifestURL != "" {
```
What is the rationale behind forbidding this? Almost nobody will ever try this, I think. But for tests, it could be handy to combine `--manifest-url` and `--archive`.

rewanth1997 replied:
As recommended by @ahmetb here, https://github.com/kubernetes-sigs/krew/pull/350#discussion_r337132131 I added this check. Please advise on how to proceed further.

corneliusweig replied:
I see, sorry for the confusion. Given that Ahmet had the same idea here https://github.com/kubernetes-sigs/krew/pull/350#discussion_r353975232, I think that we should go for the other option and only treat `--manifest` and `--manifest-url` mutually exclusive, but not `--manifest-url` and `--archive`.

rewanth1997 replied:
Done

corneliusweig commented on cmd/krew/cmd/install.go:
```diff
@@ -62,6 +62,14 @@ Remarks:
+				return errors.New("--archive cannot be specified with --manifest-url")
+			}
+
+			if *manifest != "" && *manifestURL != "" {
```
This on the other hand makes perfect sense to me.

corneliusweig commented on cmd/krew/cmd/install.go:
```diff
@@ -94,11 +102,27 @@ Remarks:
+
 				if err := validation.ValidatePlugin(plugin.Name, plugin); err != nil {
 					return errors.Wrap(err, "plugin manifest validation error")
 				}
```
This block should be extracted into its own function, so that the abstraction level in this function is consistent with lines 98,99.

rewanth1997 replied:
Done

rewanth1997 commented:
@ahmetb @corneliusweig 

All the tests are successful on my local system. Just to confirm, I forked my current working repository (https://github.com/rewanth1997/krew/tree/manifest-url) into my test account (https://github.com/testinguser883/krew/tree/manifest-url) and the build is successful there also (https://travis-ci.org/testinguser883/krew/builds/622366453). I'm not sure why travis is failing here.

I made all the suggested changes expect `--archive` and `--manifest-url` conflict.

I added the abstraction layer, removed hard-coded port number, closed response object after use and fixed a few linter bugs as well. Please review the latest code.

corneliusweig commented:
@rewanth1997 I think you need to rebase your current branch on top of `origin/master`. Here is what you need to do:
```
git remote add krew-origin git@github.com:kubernetes-sigs/krew.git
git fetch krew-origin
git rebase krew-origin/master
```

I guess you will have to resolve a few merge conflicts.

corneliusweig commented on cmd/krew/cmd/internal/read_plugin.go:
```diff
@@ -0,0 +1,60 @@
+		}
+	}
+
+	return plugin, nil
```
Why not simply
```go
if manifest != "" {
	return indexscanner.ReadPluginFile(manifest)
}
return readRemotePluginManifest(manifestURL)
```

If both are given, `manifest` should take precedence anyway.

rewanth1997 replied:
Yeah, true. Both won't be accepted because I added a pre-check to make sure `manifest` and `manifest-url` are mutually exclusive.

rewanth1997 replied:
Done

corneliusweig commented on cmd/krew/cmd/internal/read_plugin.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+// GetPlugin is an abstraction layer for manifest and manifest-url handlers and returns plugin object
+func GetPlugin(manifest string, manifestURL string) (index.Plugin, error) {
```
One of the major points of extracting logic into this function is to add unit tests. Can you add some? It should be rather easy, because you already did the major setup work for the integration tests.
In the unit tests you can also cover some error cases, for example when nothing is found for the given URL.

rewanth1997 replied:
Added multiple cases with empty file, invalid file, not existing url, etc.

corneliusweig commented on cmd/krew/cmd/install.go:
```diff
@@ -94,11 +102,14 @@ Remarks:
 					return errors.Wrap(err, "failed to load custom manifest file")
 				}
+
 				if err := validation.ValidatePlugin(plugin.Name, plugin); err != nil {
```
I would have moved that into `GetPlugin` as well, but I guess it's ok.

rewanth1997 replied:
I thought writing these separately will be able to better differentiate between `loading manifest` and `validation of manifest` errors

rewanth1997 commented on cmd/krew/cmd/internal/read_plugin.go:
```diff
@@ -0,0 +1,60 @@
+	"sigs.k8s.io/krew/pkg/index/indexscanner"
+)
+
+// readRemotePluginManifest function takes URL as input and returns plugin as output
```
Done

rewanth1997 commented on cmd/krew/cmd/internal/read_plugin.go:
```diff
@@ -0,0 +1,60 @@
+
+	plugin, err := indexscanner.DecodePluginFile(resp.Body)
+	if err != nil {
+		return index.Plugin{}, err
```
Done

rewanth1997 commented on cmd/krew/cmd/internal/read_plugin.go:
```diff
@@ -0,0 +1,60 @@
+func readRemotePluginManifest(url string) (index.Plugin, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return index.Plugin{}, err
```
Done

rewanth1997 commented:
@corneliusweig All recommended changes updated.

rewanth1997 commented:
@corneliusweig Even the mutually exclusive condition is fixed in the latest commit. All tests are passed locally. Please let me know if any further changes are required.

ahmetb commented:
The tests are failing with a [very unusual error](https://user-images.githubusercontent.com/159209/70401632-6d4a1c00-19e5-11ea-958c-591e791f752c.png).
I'll make another review or push commits directly to your branch, thanks for staying on top of this @rewanth1997.

ahmetb commented on integration_test/install_test.go:
```diff
@@ -105,6 +109,115 @@ func TestKrewInstall_Manifest(t *testing.T) {
+	}
+	server.Listener = listener
+	server.Start()
+	return server, server.URL
```
You can probably just do this:

```
httptest.NewServer(..)
return server.URL, server.Close
```

it does everything you want, and then you can embed it to the test because it's 2 lines.

ahmetb commented on integration_test/install_test.go:
```diff
@@ -105,6 +109,115 @@ func TestKrewInstall_Manifest(t *testing.T) {
+	if err == nil {
+		t.Errorf("expected install to fail but was successful")
+	}
+}
```
what does this really test?
our validation method? aren't those already covered with unit tests?

ahmetb commented on integration_test/install_test.go:
```diff
@@ -105,6 +109,115 @@ func TestKrewInstall_Manifest(t *testing.T) {
+	if err == nil {
+		t.Errorf("expected install to fail but was successful")
+	}
+}
```
what does this really test?
our validation method? aren't those already covered with unit tests?

ahmetb commented on cmd/krew/cmd/internal/read_plugin.go:
```diff
@@ -0,0 +1,49 @@
+	}
+
+	return readRemotePluginManifest(manifestURL)
+}
```
This isn't really a good way to do things.
We could do this at the call site, now you just embedded the decisionmaking here, and it doesn't have unit tests.

ahmetb commented on cmd/krew/cmd/internal/read_plugin.go:
```diff
@@ -0,0 +1,49 @@
+	}
+
+	return plugin, nil
+}
```
this doesn't have unit tests :)
would be good to unit-test this, rather than trying to test /empty.yaml /invalid.yaml etc.

ahmetb commented on cmd/krew/cmd/internal/read_plugin.go:
```diff
@@ -0,0 +1,49 @@
+// Copyright 2019 The Kubernetes Authors.
```
this cmd/ package does nearly nothing noteworthy that requires creating a new package for. especially because cmd/ directory directly has "tool" names (krew, generate-plugin-list, ...) this doesn't fit here.

ahmetb commented:
I think this PR has showed us fundamentally how we should approach accepting patches to krew codebase.

To me, minimalism is key, and ideally we should not require ~200 lines of _new_ code to add one flag like this.

I'll be preparing a PR that does the same thing. I think some issues in the PR stems from the fact that our indexscanner API isn't the best it could be. Let's see how much new code it'll end up in.

ahmetb commented on cmd/krew/cmd/install.go:
```diff
@@ -81,8 +85,8 @@ Remarks:
 				return errors.New("must specify either specify either plugin names (via positional arguments or STDIN), or --manifest; not both")
```
you missed this part. this ought to also factor in --manifest-url.

ahmetb commented:
/hold
In about an hour I prepared a patch that does this: #429 
It only results in ~120 lines of net added code while addressing the underlying technical debt and adding some missing tests.

rewanth1997 commented:
> The tests are failing with a [very unusual error](https://user-images.githubusercontent.com/159209/70401632-6d4a1c00-19e5-11ea-958c-591e791f752c.png).

Yes, even I'm not sure why? This branch is building successfully without any errors
https://travis-ci.org/testinguser883/krew/builds/622405722?utm_source=github_status&utm_medium=notification

It would be great if you can point out why the build is failing here (maybe it's because this branch is not synced with the current master repo?)

ahmetb commented:
Yeah, most likely you’ll need to rebase. 
On the other hand, we can continue on my PR as we’re getting to ~100 comments on this and we still have some way to go. I think my patch addresses this more cleanly, but thank you for doing the work on this.

rewanth1997 commented:
Yeah, sure. We will work on your PR. I will try to give inputs to the best of my knowledge.

@corneliusweig @ahmetb Thanks for considering my idea, being so patient and reviewing my PR.

corneliusweig commented:
Hey @rewanth1997, sorry that your PR doesn't make it into krew's codebase. Don't be upset though. The final draft was already much more mature than the initial one, so you probably could learn a lot by doing it. For your future work, you should focus on testing your code. You can learn a lot from looking at @ahmetb's version :)

/close
